pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,889 @@
+/*
+Copyright 2019 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package edgehub
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/pkg/errors"
+
+	"github.com/kubeedge/kubeedge/beehive/pkg/core"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/mocks/edgehub"
+	_ "github.com/kubeedge/kubeedge/pkg/devicetwin"
+	"github.com/kubeedge/kubeedge/pkg/edgehub/config"
+)
+
+//TestServer is a fake http server created for testing
+var TestServer *httptest.Server
+
+// mockAdapter is a mocked adapter implementation
+var mockAdapter *edgehub.MockAdapter
+
+//init() starts the test server and generates test certificates for testing
+func init() {
+	newTestServer()
+	generateTestCertificate()
+}
+
+// initMocks is function to initialize mocks
+func initMocks(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockAdapter = edgehub.NewMockAdapter(mockCtrl)
+}
+
+//newTestServer() starts a fake server for testing
+func newTestServer() {
+	flag := true
+	TestServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch {
+		case strings.Contains(r.RequestURI, "/proper_request"):
+			w.Write([]byte("ws://127.0.0.1:20000"))
+		case strings.Contains(r.RequestURI, "/bad_request"):
+			w.WriteHeader(http.StatusBadRequest)
+		case strings.Contains(r.RequestURI, "/wrong_url"):
+			if flag {
+				w.WriteHeader(http.StatusNotFound)
+				flag = false
+			} else {
+				w.Write([]byte("ws://127.0.0.1:20000"))
+			}
+		}
+	}))
+}
+
+//generateTestCertificate() generates fake certificates and stores them at a temporary location (/tmp/kubeedge/testData)
+func generateTestCertificate() {
```

Comment:
this function can be moved to a util file. the test code can simply make utility func calls and thus remain more cleaner.